enlighten/measurement/Measurement.py

```diff
@@ -445,18 +446,55 @@ def expand_template(self, template):
         attribute in wasatch.EEPROM, wasatch.SpectrometerState, or Measurement
         metadata (any field supported by Measurement.get_metadata). As a 
         convenience some hardcoded macros are also supported, such as {time}.
+
+        @todo We can't easily expand the set of objects whose attributes can
+              be used (like BatchCollection) without running into potential
+              namespace collisions (different objects can have identically-
+              named attributes, leading to ambiguity).  Templates should
+              move toward a prefixed notation like "m.measurement_id",
+              "b.current_batch_count" etc.  This could then be implemented
+              within a call to string.format(), giving users access to
+              full precision controls etc.
+
+              Also I'd pull this out into a TemplateFeature.
         """
         log.debug(f"expand_template: starting with template {template}")
         while True:
-            m = re.search(r"{([a-z0-9_]+)}", template, re.IGNORECASE)
+            m = re.search(r"{([a-z0-9_ ]+)}", template, re.IGNORECASE)
             if m is None:
                 return template
 
             macro = m.group(1)
             value = None
+            fmt = None
+
+            ####################################################################
+            # macro-only fields (don't map to existing data)
+            ####################################################################
+
+            if   macro == "time": value = self.timestamp.strftime("%H_%M_%S")
+            elif macro == "date": value = self.timestamp.strftime("%Y-%m-%d")
+            elif macro == "file_timestamp": value = self.timestamp.strftime("%Y-%m-%d_%H_%M_%S%f")
+
+            # note all date components are upper-case and times are lower-case 
+            # for consistency (may confuse C programmers, should make sense to 
+            # spectroscopists)
+            elif macro == "YYYY": value = self.timestamp.strftime("%Y")
+            elif macro == "MM": value = self.timestamp.strftime("%m")
+            elif macro == "DD": value = self.timestamp.strftime("%d")
+            elif macro == "hh": value = self.timestamp.strftime("%H")
+            elif macro == "mm": value = self.timestamp.strftime("%M")
+            elif macro == "ss": value = self.timestamp.strftime("%S")
+            elif macro == "ffffff": value = self.timestamp.strftime("%f")
```

Comment:
This could be an anonymous dictionary expression:

```python
value = self.timestamp.strftime({
    "time": "%H_%M_%S",
    "date": "%Y-%m-%d",
    ...
    }.get(macro))
```

Pros:
* easier to read
* easier to add entries
* less prone to hard to spot logic errors: (elif->if, value=->value==)

Cons:
* more to difficult to understand
* last elif containing `fmt =` will need to be patch at end
* this format is not as flexible to arbitrary semantic changes